src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,425 @@
+            using var disposalRegistration = _disposalSource.Token.Register(() => TrySetStream(StreamStateDisposed, value: null));
+
+            RegisterHandler(runtimeId, (Guid id, ref Stream cachedStream) =>
+            {
+                if (id != runtimeId)
+                {
+                    return false;
+                }
+
+                if (TrySetStream(StreamStateComplete, cachedStream))
+                {
+                    cachedStream = null;
+                }
+
+                // Regardless of the registrant previously waiting or cancelled,
+                // the handler should be removed from consideration.
+                return true;
+            });
+
+            cancellationSource.CancelAfter(timeout);
+            streamEvent.WaitOne();
+
+            if (StreamStateCancelled == streamState)
+            {
+                throw new TimeoutException();
+            }
+
+            if (StreamStateDisposed == streamState)
+            {
+                throw new ObjectDisposedException(nameof(ReversedDiagnosticsServer));
+            }
+
+            return stream;
+        }
+
+        internal async Task WaitForConnectionAsync(Guid runtimeId, CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            var hasConnectedStreamSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            using var methodRegistration = token.Register(() => hasConnectedStreamSource.TrySetCanceled(token));
+            using var disposalRegistration = _disposalSource.Token.Register(
+                () => hasConnectedStreamSource.TrySetException(new ObjectDisposedException(nameof(ReversedDiagnosticsServer))));
+
+            RegisterHandler(runtimeId, (Guid id, ref Stream cachedStream) =>
+            {
+                if (runtimeId != id)
+                {
+                    return false;
+                }
+
+                // Check if the registrant was already finished.
+                if (hasConnectedStreamSource.Task.IsCompleted)
+                {
+                    return true;
+                }
+
+                if (!TestStream(cachedStream))
+                {
+                    cachedStream.Dispose();
+                    cachedStream = null;
+                    return false;
+                }
+
+                // Found a stream that is valid; signal completion if possible.
+                hasConnectedStreamSource.TrySetResult(true);
+
+                // Regardless of the registrant previously waiting or cancelled,
+                // the handler should be removed from consideration.
+                return true;
+            });
+            
+            try
+            {
+                // Wait for the handler to verify we have a connected stream
+                await hasConnectedStreamSource.Task.ConfigureAwait(false);
```

Comment:
WaitForConnectionAsync should not have a try...catch around the Task here so that all exceptions cause the method's Task to be set to the Exception or Cancelled states.

Reply:
Fixed